src/modules/accounting/accounting.py

```diff
@@ -0,0 +1,239 @@
+        """
+        Calculate exited validators count in all modules.
+        Exclude modules without changes from the report.
+        """
+        stacking_modules = self.w3.lido_validators.get_staking_modules(blockstamp)
+
+        exited_validators = self.lido_validator_state_service.get_exited_lido_validators(blockstamp)
+
+        module_stats: dict[int, int] = defaultdict(int)
+
+        for (module_id, _), validators_exited_count in exited_validators.items():
+            module_stats[module_id] += validators_exited_count
+
+        for module in stacking_modules:
+            if module_stats[module.id] == module.exited_validators_count:
+                del module_stats[module.id]
+
+        return list(module_stats.keys()), list(module_stats.values())
+
+    @lru_cache(maxsize=1)
+    def _get_consensus_lido_state(self, blockstamp: ReferenceBlockStamp) -> tuple[int, Gwei]:
+        lido_validators = self.w3.lido_validators.get_lido_validators(blockstamp)
+
+        count = len(lido_validators)
+        total_balance = Gwei(sum(int(validator.balance) for validator in lido_validators))
+
+        logger.info({'msg': 'Calculate consensus lido state.', 'value': (count, total_balance)})
+        return count, total_balance
+
+    def _get_last_withdrawal_request_to_finalize(self, blockstamp: ReferenceBlockStamp) -> int:
+        chain_config = self.get_chain_config(blockstamp)
+        frame_config = self.get_frame_config(blockstamp)
+
+        is_bunker = self._is_bunker(blockstamp)
+        withdrawal_vault_balance = self.w3.lido_contracts.get_withdrawal_balance(blockstamp)
+        el_rewards_vault_balance = self.w3.lido_contracts.get_el_vault_balance(blockstamp)
+        finalization_share_rate = self._get_finalization_shares_rate(blockstamp)
+
+        withdrawal_service = Withdrawal(self.w3, blockstamp, chain_config, frame_config)
+
+        last_wr_id = withdrawal_service.get_next_last_finalizable_id(
+            is_bunker,
+            finalization_share_rate,
+            withdrawal_vault_balance,
+            el_rewards_vault_balance,
+        )
+        logger.info({'msg': 'Calculate last withdrawal id to finalize.', 'value': last_wr_id})
+        return last_wr_id
+
+    @lru_cache(maxsize=1)
+    def _get_finalization_shares_rate(self, blockstamp: ReferenceBlockStamp) -> int:
+        simulation = self.get_rebase_after_report(blockstamp)
+        shares_rate = simulation.post_total_pooled_ether * SHARE_RATE_PRECISION_E27 // simulation.post_total_shares
+        logger.info({'msg': 'Calculate shares rate.', 'value': shares_rate})
+        return shares_rate
+
+    def get_rebase_after_report(self, blockstamp: ReferenceBlockStamp, cl_only=False) -> LidoReportRebase:
+        chain_conf = self.get_chain_config(blockstamp)
+        frame_config = self.get_frame_config(blockstamp)
+
+        last_ref_slot = self.report_contract.functions.getLastProcessingRefSlot().call(
```

Comment:
https://github.com/lidofinance/lido-oracle/pull/213#discussion_r1118041474

Reply:
do we want a sanity check here for last_ref_slot? assert (last_ref_slot < blockstamp.…